lib/spack/spack/filesystem_view.py

```diff
@@ -0,0 +1,516 @@
+                    extensions_layout=self.extensions_layout)
+
+        except ExtensionAlreadyInstalledError:
+            # As we use sets in add_specs(), the order in which packages get
+            # activated is essentially random. So this spec might have already
+            # been activated as dependency of another package -> fail silently
+            pass
+
+        # make sure the meta folder is linked as well (this is not done by the
+        # extension-activation mechnism)
+        if not self.check_added(spec):
+            self.link_meta_folder(spec)
+
+        return True
+
+    def add_standalone(self, spec):
+        if spec.package.is_extension:
+            tty.error(self._croot + 'Package %s is an extension.'
+                      % spec.name)
+            return False
+
+        if self.check_added(spec):
+            tty.warn(self._croot + 'Skipping already linked package: %s'
+                     % colorize_spec(spec))
+            return True
+
+        tree = LinkTree(spec.prefix)
+
+        if not self.ignore_conflicts:
+            conflict = tree.find_conflict(self.root)
+            if conflict is not None:
+                tty.error(self._croot +
+                          "Cannot link package %s, file already exists: %s"
+                          % (spec.name, conflict))
+                return False
+
+        conflicts = tree.merge(self.root, link=self.link,
+                               ignore=ignore_metadata_dir,
+                               ignore_conflicts=self.ignore_conflicts)
+        self.link_meta_folder(spec)
+
+        if self.ignore_conflicts:
+            for c in conflicts:
+                tty.warn(self._croot + "Could not link: %s" % c)
+
+        if self.verbose:
+            tty.info(self._croot + 'Linked package: %s' % colorize_spec(spec))
+        return True
+
+    def check_added(self, spec):
+        assert spec.concrete
+        return spec == self.get_spec(spec)
+
+    def remove_specs(self, *specs, **kwargs):
+        assert all((s.concrete for s in specs))
+        with_dependents = kwargs.get("with_dependents", True)
+        with_dependencies = kwargs.get("with_dependencies", False)
+
+        specs = set(specs)
+
+        if with_dependencies:
+            specs = get_dependencies(specs)
+
+        if kwargs.get("exclude", None):
+            specs = set(filter_exclude(specs, kwargs["exclude"]))
+
+        all_specs = set(self.get_all_specs())
+
+        to_deactivate = specs
+        to_keep = all_specs - to_deactivate
+
+        dependents = find_dependents(to_keep, to_deactivate)
+
+        if with_dependents:
+            # remove all packages depending on the ones to remove
+            if len(dependents) > 0:
+                tty.warn(self._croot +
+                         "The following dependents will be removed: %s"
+                         % ", ".join((s.name for s in dependents)))
+                to_deactivate.update(dependents)
```

Comment:
Since you go through the work of aggregating dependents here, my understanding is that if you sorted them based on a post-order traversal of the dependency DAG that calling ```Package.do_deactivate``` with ```with_dependents=True``` would not be required. I don't think there's any harm in doing it this way although the redundancy may be confusing. If I'm right I think it would make sense to make a comment on it (I don't necessarily think it requires a change).